### DISCUSSION TITLE
Multiple widgets listening to the same Provider rebuilds with error

### DISCUSSION THREAD

**Discussion by mihael998** [rrousselGit/riverpod#1211] [Q&A]
Hi, this is my situation:
I have one Provider (StateNotifierProvider) which holds and notifies a state that contains a List<Data>. Let's say the provider is called `listProvider`.
I have a `ListWidget` which watches this provider and builds a list of `ItemWidget` based on the data that `listProvider` emits! (To be more precise actually the ListWidget only watches changes to the length of the list for optimization reasons).
Every single `ItemWidget`, to be able to rebuild based on a single Data instance (which was emitted from `listProvider`) , watches a provider called `itemProvider` (which uses the family modifier). The provider `itemProvider` to be able to return the `Data` object based on the Id, has to watch `listProvider`.
Here's a schema of the realtionships:
![Providers](https://user-images.githubusercontent.com/10122310/154794943-c1f87ea0-14dd-4a9f-af81-e246ee02d5d0.jpeg)

Here's my question:
Whenever I remove an element from listProvider, the `ListWidget` should rebuild. Before the `ListWidget` rebuilds, every single `itemProvider` reruns because the `listProvider` has been updated. The problem is that `itemProvider` with the ID of the element that has just been removed from the list, still exists and still listens to the changes. So in the body of the `itemProvider` an error throws because it tries to read an element from the list returned from `listProvider` which does not exist anymore. 

Is there a way to manage this kind of situations? What do you suggest?

**Comment by rrousselGit** [comment]
Without code, the problem is unclear.

This issue likely wouldn't happen with proper separation of concern. Chances are itemprovider and listProvider are tightly coupled in a way that they shouldn't


But if don't want to re-architecture your code, you can use `ref.listen` instead of `ref.watch` within `itemProvider` to manually handle updates to handle those cases.
